diff -urN tn5250-0.17.4.org/configure.ac tn5250-0.17.4/configure.ac
--- tn5250-0.17.4.org/configure.ac	2008-11-21 01:53:32.000000000 +0100
+++ tn5250-0.17.4/configure.ac	2018-09-22 20:12:20.298123448 +0200
@@ -152,13 +152,13 @@
 dnl **
 if test -n $sslincludedir; then
      CPPFLAGS="$CPPFLAGS $sslincludedir"
-     AC_CHECK_LIB(crypto,CRYPTO_num_locks)
-     if test "$ac_cv_lib_crypto_CRYPTO_num_locks" != "yes"
+     AC_CHECK_LIB(crypto,CRYPTO_free)
+     if test "$ac_cv_lib_crypto_CRYPTO_free" != "yes"
      then
              AC_MSG_ERROR([** Unable to find OpenSSL libraries!])
      fi
-     AC_CHECK_LIB(ssl,SSL_library_init)
-     if test "$ac_cv_lib_ssl_SSL_library_init" != "yes"
+     AC_CHECK_LIB(ssl,SSL_new)
+     if test "$ac_cv_lib_ssl_SSL_new" != "yes"
      then
              AC_MSG_ERROR([** Unable to find OpenSSL libraries!])
      fi
diff -urN tn5250-0.17.4.org/lib5250/sslstream.c tn5250-0.17.4/lib5250/sslstream.c
--- tn5250-0.17.4.org/lib5250/sslstream.c	2008-11-21 09:12:21.000000000 +0100
+++ tn5250-0.17.4/lib5250/sslstream.c	2018-09-22 20:13:26.296754215 +0200
@@ -368,16 +368,8 @@
         methstr[4] = '\0';
    }
 
-   if (!strcmp(methstr, "ssl2")) {
-        meth = SSLv2_client_method();         
-        TN5250_LOG(("SSL Method = SSLv2_client_method()\n"));
-   } else if (!strcmp(methstr, "ssl3")) {
-        meth = SSLv3_client_method();         
-        TN5250_LOG(("SSL Method = SSLv3_client_method()\n"));
-   } else {
-        meth = SSLv23_client_method();         
-        TN5250_LOG(("SSL Method = SSLv23_client_method()\n"));
-   }
+   meth = SSLv23_client_method();         
+   TN5250_LOG(("SSL Method = SSLv23_client_method()\n"));
 
 /*  create a new SSL context */
 
